b9lab-content/3-main-concepts/02-architecture.md

```diff
@@ -1 +1,294 @@
+
+Notice how our application also has to handle the list of validators sent by Tendermint. Let's gloss over this _detail_.
+
+#### A serialized transaction.
+
+We need to decide how to represent a move. In our ready-made implementation, a position `Pos` is represented by [2 `int`](https://github.com/batkinson/checkers-go/blob/a09daeb/checkers/checkers.go#L42-L45), and a move by [2 `Pos`](https://github.com/batkinson/checkers-go/blob/a09daeb/checkers/checkers.go#L168). We can decide to represent a serialized move as 4 `int`, the first 2 for the original position `src`, and the next 2 for the destination position `dst`.
+
+#### `BeginBlock`: a new block is about to be created
+
+[This command](https://github.com/tendermint/spec/blob/c939e15/spec/abci/abci.md#beginblock) instructs the application to load its state at the right place. More precisely, inside `header: Header`, as per its [detailed definition](https://github.com/tendermint/spec/blob/c939e15/spec/core/data_structures.md#header), we find:
+
+> `AppHash: []byte`: Arbitrary byte array returned by the application after executing and commiting the previous block. It serves as the basis for validating any merkle proofs that comes from the ABCI application and represents the state of the actual application rather than the state of the blockchain itself. The first block's `block.Header.AppHash` is given by `ResponseInitChain.app_hash`.
+
+This instructs the application to load, from its database, the right state of the blockchain, which includes the correct `/store/board`. It is important to insist on the fact that the application needs to be able to load a known state at any _point in time_. Indeed, there could have been a crash or a restore of some sort that has desynchronized Tendermint and the application.
+
+In case the header has omitted the `AppHash`, which should never happen, the application should work off the last state it has arrived at.
+
+With this state loaded, the application is ready to respond to the upcoming `CheckTx` and `DeliverTx`.
+
+#### `CheckTx`: a new transaction appears in the transaction pool
+
+That's where Tendermint [asks](https://github.com/tendermint/spec/blob/c939e15/spec/abci/abci.md#checktx) our application whether the transaction is worth keeping at all. Because we want to simplify to the maximum, we only concern ourselves with whether there is a valid move in the transaction.
+
+For that, we check whether there are 4 `int` in the serialized information. We can also check that the `int` themselves are within the boundaries of the board, i.e. between `0` and `7`.
+
+It is better _not_ to check whether we have a [valid move](https://github.com/batkinson/checkers-go/blob/a09daeb/checkers/checkers.go#L168):
+
+> `func (game *Game) ValidMove(src, dst Pos) bool`
+
+As a transaction may contain a move that is valid only when actually delivered, for instance if it needs another move to take place before it.
```

Comment:
This was unclear to me when I read it the first time, especially in the context of checkers, because at this stage of the exercise you’d probably expect only 1 move to occur per block and so it wouldn’t be entirely clear why another move might take place before it or why we would have to wait for anything.

Feels to me this is also related to efficiency as we don't want to perform a full check on every transaction in the mempool? (right?)